src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoInternalOfT.cs

```diff
@@ -93,5 +124,28 @@ private void SetCreateObjectFunc(Func<T>? createObjectFunc)
                 CreateObject = () => createObjectFunc();
             }
         }
+
+        private void DetermineIfCanUseSerializeFastPath()
+        {
+            Debug.Assert(Options == Options._context!.Options);
+            JsonSerializerOptions? contextDefaultOptions = Options._context!.DefaultOptions;
+            if (contextDefaultOptions == null)
+            {
+                throw new InvalidOperationException($"To specify a fast-path implementation for type {typeof(T)}, context {Options._context.GetType()} must specify default options.");
+            }
+
+            UseFastPathOnWrite =
```

Comment:
Are custom converters supported for fast path and if not, how is that checked?

Reply:
We have a design-time option specifying if metadata initialization code-paths can ignore runtime-provided custom converters (i.e. assume none will be added), but that's a separate concern.

> Are custom converters supported for fast path 

Not today, but we could support it in the future. From the example in the description:

```cs
private static void JsonMessageSerialize(global::System.Text.Json.Utf8JsonWriter writer, global::System.Text.Json.SourceGeneration.Tests.JsonMessage value)
{
    // New logic starts
    var options = JsonContext.Default.Options;
    JsonConverter<T>? converter = GetRuntimeConverter(options); // store and cache converter for type.
    if (converter != null)
    {
        converter.Write(writer, value, options);
        return;
    }
    // New logic ends

    if (value == null)
    {
        writer.WriteNullValue();
        return;
    }

    writer.WriteStartObject();

    if (value.Message != null)
    {
        writer.WriteString(messagePropName, value.Message);
    }

    writer.WriteEndObject();
}
```

> how is that checked?

I'll add a new condition on the next line: `Options.Converters.Count == 0` (or something else that aligns with feedback from https://github.com/dotnet/runtime/pull/53212#discussion_r639108481) and a test to make sure fast-path isn't called when there's a converter.